docs/cluster/networking.md

```diff
@@ -0,0 +1,142 @@
+
+Before discussing the Kubernetes approach to networking, it is worthwhile to
+review the "normal" way that networking works with Docker.  By default, Docker
+uses host-private networking.  It creates a virtual bridge, called `docker0` by
+default, and allocates an [RFC1918](https://tools.ietf.org/html/rfc1918) range
+for that bridge.  Each container that Docker creates is allocated a virtual
+ethernet device (called `veth`).  The other end of the veth is connected to the
+bridge.  The in-container interface is given an IP address from the bridge's
+address range.
+
+The result is that Docker containers can talk to other containers only if they
+are on the same machine (and thus the same virtual bridge).  Containers on
+different machines can not reach each other - in fact they may end up with the
+exact same network ranges and IP addresses.
+
+In order for Docker containers to communicate across nodes, they must be
+allocated ports on the machine's own IP address, which are then forwarded or
+proxied to the containers.  This obviously means that containers must either
+coordinate which ports they use very carefully or else be allocated ports
+dynamically.
+
+## Kubernetes model
+
+Coordinating ports across multiple developers is very difficult and runs against
+the general mindset of Kubernetes.  Dynamic port allocation brings a lot of
+complications to the system - every application has to take ports as flags,
+the API servers have to know how to insert dynamic port numbers into
+configuration blocks, services have to know how to find each other, etc.  Rather
+than deal with this, Kubernetes takes a different approach.
+
+The fundamental requirements that Kubernetes makes about networking are:
+   * all containers can communicate with all other containers without NAT
+   * all nodes can communicate with all containers (and vice-versa) without NAT
+   * the IP that a container sees itself as is the same IP that others see it as
+
+What this means in practice is that you can not just take two computers
+running Docker and expect kubernetes to work.  You must ensure that the
+fundamental requirements are met.
+
+This model is not only less complex overall, but it is principally compatible
+with the desire for Kubernetes to enable low-friction migration from VMs.  If
+your job previously ran in a VM, your VM had an IP and could talk to other VMs
+in your project.  This is the same basic model.
+
+Until now this document has talked about containers.  In reality, Kubernetes
+applies IP addresses at the `Pod` scope - containers within a `Pod` share their
+network namespaces - including their IP address.  This means that containers
+within a `Pod` can all reach each other’s ports on `localhost`.  This does imply
+that containers within a `Pod` must coordinate port usage, but this is no
+different that processes in a VM.  We call this the "IP-per-pod" model.  It is
+possible to request host ports, but this is reduced to a very niche operation.
```

Comment:
When you request a hostPort, does the Pod see that it has two IP addresses, or does it just have the hostIP, or just the  PodIP, with some magic routing of traffic just on one port to the hostIP?